From: csagan5 <32685696+csagan5@users.noreply.github.com>
Date: Thu, 22 Jul 2021 23:10:52 +0200
Subject: Client hints overrides

Disable critical client hints and network early hints
Use Google Chrome branding for client hints

License: GPL-3.0-only - https://spdx.org/licenses/GPL-3.0-only.html
---
--- a/components/client_hints/browser/client_hints.cc
+++ b/components/client_hints/browser/client_hints.cc
@@ -36,7 +36,7 @@ ParseInitializeClientHintsStorage() {
   auto results =
       base::flat_map<url::Origin,
                      std::vector<network::mojom::WebClientHintsType>>();
-
+  if ((true)) return results;
   std::string raw_client_hint_json =
       base::CommandLine::ForCurrentProcess()->GetSwitchValueASCII(
           switches::kInitializeClientHintsStorage);
@@ -201,7 +201,8 @@ void ClientHints::PersistClientHints(
   // TODO(tbansal): crbug.com/735518. Disable updates to client hints settings
   // when cookies are disabled for |primary_origin|.
   content_settings::ContentSettingConstraints constraints;
-  constraints.set_session_model(content_settings::mojom::SessionModel::DURABLE);
+  constraints.set_session_model(content_settings::mojom::SessionModel::USER_SESSION);
+  constraints.set_lifetime(base::Minutes(1));
   settings_map_->SetWebsiteSettingDefaultScope(
       primary_url, GURL(), ContentSettingsType::CLIENT_HINTS,
       base::Value(std::move(client_hints_dictionary)), constraints);
--- a/content/browser/client_hints/client_hints.cc
+++ b/content/browser/client_hints/client_hints.cc
@@ -57,6 +57,7 @@
 #include "third_party/blink/public/common/features.h"
 #include "third_party/blink/public/common/page/page_zoom.h"
 #include "third_party/blink/public/common/user_agent/user_agent_metadata.h"
+#include "third_party/blink/renderer/platform/runtime_enabled_features.h"
 #include "ui/display/display.h"
 #include "ui/display/screen.h"
 #include "url/origin.h"
@@ -601,10 +602,21 @@ bool IsClientHintEnabled(const ClientHin
 
 bool IsClientHintAllowed(const ClientHintsExtendedData& data,
                          WebClientHintsType type) {
+  if (!blink::RuntimeEnabledFeatures::UserAgentClientHintEnabled())
+    return false;
+
+  bool allowed = false;
+  if (type == WebClientHintsType::kUA ||
+      type == WebClientHintsType::kUAMobile ||
+      type == WebClientHintsType::kUAPlatform ||
+      type == WebClientHintsType::kUAModel ||
+      type == WebClientHintsType::kUAPlatformVersion) {
+    allowed = true;
+  }
   if (data.is_outermost_main_frame) {
-    return true;
+    return allowed;
   }
-  return (data.permissions_policy->IsFeatureEnabledForOrigin(
+  return ((false)) && (data.permissions_policy->IsFeatureEnabledForOrigin(
       network::GetClientHintToPolicyFeatureMap().at(type),
       data.resource_origin));
 }
@@ -629,6 +641,7 @@ bool IsJavascriptEnabled(FrameTreeNode*
 void UpdateIFramePermissionsPolicyWithDelegationSupportForClientHints(
     ClientHintsExtendedData& data,
     const network::ParsedPermissionsPolicy& container_policy) {
+  if ((true)) return;
   if (container_policy.empty()) {
     return;
   }
--- a/net/http/http_network_transaction.cc
+++ b/net/http/http_network_transaction.cc
@@ -126,6 +126,7 @@ const size_t kMaxRestarts = 32;
 
 // Returns true when Early Hints are allowed on the given protocol.
 bool EarlyHintsAreAllowedOn(HttpConnectionInfo connection_info) {
+  if ((true)) return false;
   switch (connection_info) {
     case HttpConnectionInfo::kHTTP0_9:
     case HttpConnectionInfo::kHTTP1_0:
--- a/services/network/public/cpp/client_hints.cc
+++ b/services/network/public/cpp/client_hints.cc
@@ -111,7 +111,7 @@ const DecodeMap& GetDecodeMap() {
 }  // namespace
 
 std::optional<std::vector<network::mojom::WebClientHintsType>>
-ParseClientHintsHeader(const std::string& header) {
+ParseClientHintsHeader(const std::string& header) { // Parse Client Hints Header
   // Accept-CH is an sh-list of tokens; see:
   // https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-header-structure-19#section-3.1
   std::optional<net::structured_headers::List> maybe_list =
--- a/third_party/blink/common/client_hints/client_hints.cc
+++ b/third_party/blink/common/client_hints/client_hints.cc
@@ -14,6 +14,7 @@
 #include "base/strings/string_util.h"
 #include "services/network/public/cpp/client_hints.h"
 #include "services/network/public/cpp/permissions_policy/client_hints_permissions_policy_mapping.h"
+#include "third_party/blink/public/common/features_generated.h"
 #include "services/network/public/cpp/permissions_policy/permissions_policy.h"
 #include "third_party/blink/public/common/features.h"
 #include "url/origin.h"
@@ -22,11 +23,12 @@ namespace blink {
 
 bool IsClientHintSentByDefault(network::mojom::WebClientHintsType type) {
   switch (type) {
-    case network::mojom::WebClientHintsType::kSaveData:
     case network::mojom::WebClientHintsType::kUA:
     case network::mojom::WebClientHintsType::kUAMobile:
     case network::mojom::WebClientHintsType::kUAPlatform:
-      return true;
+      return base::FeatureList::IsEnabled(blink::features::kUserAgentClientHint);
+    case network::mojom::WebClientHintsType::kSaveData:
+      return false;
     default:
       return false;
   }
--- a/third_party/blink/common/client_hints/enabled_client_hints.cc
+++ b/third_party/blink/common/client_hints/enabled_client_hints.cc
@@ -15,6 +15,14 @@ namespace {
 using ::network::mojom::WebClientHintsType;
 
 bool IsDisabledByFeature(const WebClientHintsType type) {
+  if (type == WebClientHintsType::kUA ||
+        type == WebClientHintsType::kUAMobile ||
+        type == WebClientHintsType::kUAPlatform ||
+        type == WebClientHintsType::kUAModel ||
+        type == WebClientHintsType::kUAPlatformVersion) {
+    return false;
+  }
+  if ((true)) return true;
   switch (type) {
     case WebClientHintsType::kUA:
     case WebClientHintsType::kUAArch:
--- a/third_party/blink/renderer/core/frame/navigator_ua.idl
+++ b/third_party/blink/renderer/core/frame/navigator_ua.idl
@@ -4,6 +4,8 @@
 
 // https://github.com/WICG/ua-client-hints
 
-interface mixin NavigatorUA {
+[
+    RuntimeEnabled=UserAgentClientHint
+] interface mixin NavigatorUA {
   [SecureContext] readonly attribute NavigatorUAData userAgentData;
 };
--- a/third_party/blink/renderer/core/frame/navigator_ua_data.cc
+++ b/third_party/blink/renderer/core/frame/navigator_ua_data.cc
@@ -130,6 +130,7 @@ const String& NavigatorUAData::platform(
 }
 
 bool AllowedToCollectHighEntropyValues(ExecutionContext* execution_context) {
+  if ((true)) return false;
   // To determine whether a document is allowed to get high-entropy
   // client hints returned by navigator.userAgentData.getHighEntropyValues(),
   // we need to check the "ch-ua-high-entropy-values" policy:
--- a/third_party/blink/renderer/core/frame/navigator_ua_data.idl
+++ b/third_party/blink/renderer/core/frame/navigator_ua_data.idl
@@ -4,7 +4,10 @@
 
 // https://github.com/WICG/ua-client-hints
 
-[Exposed=(Window,Worker)] interface NavigatorUAData {
+[
+    RuntimeEnabled=UserAgentClientHint,
+    Exposed=(Window,Worker)
+] interface NavigatorUAData {
   [MeasureAs=NavigatorUAData_Brands] readonly attribute FrozenArray<NavigatorUABrandVersion> brands;
   [MeasureAs=NavigatorUAData_Mobile]  readonly attribute boolean mobile;
   [MeasureAs=NavigatorUAData_Platform] readonly attribute DOMString platform;
